File: AMFDivision.java
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

public final class AMFDivision {

    public record AMFFrame(BigInteger quotient, BigInteger remainder, BigInteger divisor) {
        public BigInteger reconstruct() {
            return divisor.multiply(quotient).add(remainder);
        }

        @Override
        public String toString() {
            return "AMFFrame{q=" + quotient + ", r=" + remainder + ", b=" + divisor + "}";
        }
    }

    public static AMFFrame divide(BigInteger dividend, BigInteger divisor) {
        if (divisor.equals(BigInteger.ZERO)) {
            throw new ArithmeticException("Division by zero");
        }

        BigInteger[] qr = dividend.divideAndRemainder(divisor);

        BigInteger q = qr[0];
        BigInteger r = qr[1];

        // Normalize remainder for Euclidean form: 0 <= r < |divisor|
        if (r.signum() < 0) {
            if (divisor.signum() > 0) {
                q = q.subtract(BigInteger.ONE);
                r = r.add(divisor);
            } else {
                q = q.add(BigInteger.ONE);
                r = r.subtract(divisor);
            }
        }

        return new AMFFrame(q, r, divisor);
    }

    public static List<BigInteger> expand(BigInteger dividend, BigInteger divisor, int base, int digits) {
        if (base < 2) {
            throw new IllegalArgumentException("Base must be >= 2");
        }

        List<BigInteger> output = new ArrayList<>();
        BigInteger radix = BigInteger.valueOf(base);

        AMFFrame frame = divide(dividend, divisor);

        // Whole part
        output.add(frame.quotient());

        BigInteger remainder = frame.remainder();

        // Fractional / infinite-frame continuation
        for (int i = 0; i < digits && remainder.signum() != 0; i++) {
            remainder = remainder.multiply(radix);

            AMFFrame next = divide(remainder, divisor);

            output.add(next.quotient());
            remainder = next.remainder();
        }

        return output;
    }

    public static void main(String[] args) {
        BigInteger a = BigInteger.valueOf(7);
        BigInteger b = BigInteger.valueOf(3);

        AMFFrame frame = divide(a, b);
        System.out.println(frame);

        List<BigInteger> expansion = expand(a, b, 10, 10);
        System.out.println(expansion);
        // Output: [2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]
    }
}

File: AMFDivision.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long dividend;
    long divisor;
    long quotient;
    long remainder;
} AMFDivisionResult;

AMFDivisionResult amf_divide(long a, long b) {
    if (b == 0) {
        fprintf(stderr, "Error: division by zero\n");
        exit(EXIT_FAILURE);
    }

    long q = a / b;
    long r = a % b;

    // Normalize so remainder is always 0 <= r < |b|
    if (r < 0) {
        if (b > 0) {
            q -= 1;
            r += b;
        } else {
            q += 1;
            r -= b;
        }
    }

    AMFDivisionResult result;
    result.dividend = a;
    result.divisor = b;
    result.quotient = q;
    result.remainder = r;
    return result;
}

double amf_to_double(AMFDivisionResult result) {
    return (double)result.dividend / (double)result.divisor;
}

void print_amf_result(AMFDivisionResult result) {
    printf("AMFDivisionResult{dividend=%ld, divisor=%ld, quotient=%ld, remainder=%ld, expression=%ld + %ld/%ld}\n",
           result.dividend,
           result.divisor,
           result.quotient,
           result.remainder,
           result.quotient,
           result.remainder,
           result.divisor);
}

int main(void) {
    print_amf_result(amf_divide(10, 3));
    print_amf_result(amf_divide(3, 10));
    print_amf_result(amf_divide(-10, 3));
    print_amf_result(amf_divide(10, -3));
    print_amf_result(amf_divide(-10, -3));
    return 0;
}
